### ISSUE TITLE
Implementation of SHA in Dafny/Spartan

### ISSUE THREAD

**Issue by vkanne-msft** [mitls/mitls-fstar#108]
This task is used to track progress of implementation of SHA in Dafny/Spartan

**Comment by BarryBo** [comment]
See issue 112 for full details.  SHA256 extraction to Kremlin is implemented in Dafny, but on hold until Kremlin adds required features.

**Comment by BarryBo** [comment]
In progress.

**Comment by vkanne-msft** [comment]
Blocked on following Kremlin issues 
Current issues:
-          CopyArray<T>, as templates aren’t easy to support in Kremlin.  Jay, I sent you mail about this.  The fix should be just a couple of simple source edits to hard-code it to uint32 types.
-          BitwiseAdd32() and DafnyCastByteArrayToWordArray() are flagged as axioms in Dafny, resulting in undefined functions both in C# and C extraction.  I assume this is work-in-progress placeholder code.
-          Issue fstarlang/kremlin/#2 – checked-in kremlib.h uses a datatype that the Microsoft C compiler doesn’t support
-          Issue fstarlang/kremlin/#3 – “return” from a void function/method generates C “return (0);”
-          Issue fstarlang/kremlin/#4 – type declarations are unordered in Dafny, but C requires a dependency graph
-          Issue fstarlang/kremlin/#5 – feature request, for declaring but not defining a function, for the Dafny calls to Spartan functions
-          Issue fstarlang/kremlin/#6 – problem passing byref structs

**Comment by jaylorch** [comment]
I checked in an updated version of the SHA Dafny files that replaces CopyArray<T> with CopyUint32Array,  to avoid templates.  It's in the master branch of Spartan.
